flow/project/cli/config/processor.go

```diff
@@ -0,0 +1,82 @@
+/*
+ * Flow CLI
+ *
+ * Copyright 2019-2020 Dapper Labs, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package config
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"regexp"
+	"strings"
+)
+
+var (
+	envRegex  *regexp.Regexp = regexp.MustCompile(`\$\{env\:(.+)\}`)
+	fileRegex *regexp.Regexp = regexp.MustCompile(`\"\$\{file\:(.+)\}\"`)
+)
+
+// processEnv finds env variables and insert env values
+func processEnv(raw string) string {
+	envMatches := envRegex.FindAllStringSubmatch(raw, -1)
+
+	for _, match := range envMatches {
+		raw = strings.ReplaceAll(
+			raw,
+			match[0],
+			getEnvVariable(match[1]),
+		)
+	}
+
+	return raw
+}
+
+// processFile finds file variables and insert content
+func processFile(raw string) string {
```

Comment:
Here's one way to implement this in order to mock out the actual filesystem in tests.

[Afero](https://github.com/spf13/afero) is an abstraction system for filesystems that I think could work well in this case.

```go
import "github.com/spf13/afero"

type Preprocessor struct {
  filesystem afero.Fs
}

func NewPreprocessor(filesystem afero.Fs) *Preprocessor {
  return &Preprocessor{filesystem: filesystem}
}

func (p *Preprocessor) Run(raw string) string {
  raw = p.processEnv(raw)
  raw = p.processFile(raw)
  return raw
}

func (p *Preprocessor) processFile(raw string) string {
  // ...
  
  configFileRaw, err := p.filesystem.Open(match[1])
  
  // ...
}
```

```go
import "github.com/spf13/afero"

// in real implementation
preprocessor := NewPreprocessor(afero.NewOsFs())
// ...
```

```go
import "github.com/spf13/afero"

// in tests
mockFS := afero.NewMemMapFs()

afero.WriteFile(mockFS, "src/a/b", []byte("file b"), 0644)

preprocessor := NewPreprocessor(mockFS)
// ...
```

Reply:
awesome, I will implement this